src/blk/kernel/KernelDevice.cc

```diff
@@ -1537,23 +1546,17 @@ void KernelDevice::handle_conf_change(const ConfigProxy& conf,
         discard_threads.emplace_back(new DiscardThread(this, i));
         discard_threads.back()->create("bstore_discard");
       }
-    } else {
-      // Decrease? Signal threads after telling them to stop
+    // Decrease? Signal threads after telling them to stop
+    } else if (newval < oldval) {
       dout(10) << __func__ << " stopping " << (oldval - newval) << " existing discard threads" << dendl;
 
-      // Decreasing to zero is exactly the same as disabling async discard.
-      // Signal all threads to stop
-      if(newval == 0) {
-        _discard_stop();
-      } else {
-        // Signal the last threads to quit, and stop tracking them
-        for(uint64_t i = oldval - 1; i >= newval; i--)
-        {
-          // Also detach the thread so we no longer need to join
-          discard_threads[i]->stop = true;
-          discard_threads[i]->detach();
-          discard_threads.erase(discard_threads.begin() + i);
-        }
+      // Signal the last threads to quit, and stop tracking them
+      for(uint64_t i = oldval - 1; i >= newval && i != UINT64_MAX; i--)
+      {
+        // Also detach the thread so we no longer need to join
+        discard_threads[i]->stop = true;
+        discard_threads[i]->detach();
+        discard_threads.erase(discard_threads.begin() + i);
```

Comment:
Looks like in case of new_val == 0 this could suffer from the same "slow draining" issue as the one I highlighted in discard_stop above. Shouldn't we resize discard_threads first, and then stop the relevant threads?

Reply:
Sure! I didn't look too hard at this when I was moving it around but I thought of a way to make this all a bit cleaner. New commit pushed